httpbp/client_middlewares.go

```diff
@@ -0,0 +1,196 @@
+package httpbp
+
+import (
+	"io"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	retry "github.com/avast/retry-go"
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/reddit/baseplate.go/retrybp"
+	"github.com/reddit/baseplate.go/tracing"
+	"github.com/sony/gobreaker"
+)
+
+// ClientMiddleware is used to build HTTP client middlewares by implementing
+// http.RoundTripper which http.Client accepts as Transport.
+type ClientMiddleware func(next http.RoundTripper) http.RoundTripper
+
+type roundTripper func(req *http.Request) (*http.Response, error)
+
+func (f roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req)
+}
+
+// NewClient returns a standard HTTP client wrapped with the default
+// middlewares plus any additional client middlewares passed into this
+// function.
+func NewClient(config ClientConfig, middlewares ...ClientMiddleware) *http.Client {
+	transport := &http.Transport{
+		MaxConnsPerHost: config.MaxConnections,
+	}
+
+	defaults := []ClientMiddleware{
+		ClientErrorWrapper(),
+		MonitorClient(config),
+		Retries(),
+		MaxConcurrency(config),
+		CircuitBreaker(config),
+	}
+
+	middlewares = append(middlewares, defaults...)
+
+	return &http.Client{
+		Transport: applyMiddlewares(transport, middlewares...),
+	}
+}
+
+// ClientErrorWrapper applies ClientErrorFromResponse to the returned response
+// ensuring a HTTP status response outside the range [200, 400) is wrapped in
+// an error relieving users from the need to check the status response.
+func ClientErrorWrapper() ClientMiddleware {
+	return func(next http.RoundTripper) http.RoundTripper {
+		return roundTripper(func(req *http.Request) (resp *http.Response, err error) {
+			resp, err = next.RoundTrip(req)
+			err = ClientErrorFromResponse(resp)
```

Comment:
sorry. this means that we could return both non-nil resp and non-nil err, and users need to `defer httpbp.DrainAndClose(resp.Body)` even if err is non-nil, which can also be a footgun. for that reason, I think we should not make this a default middleware (but still provide it), and also document that this could return both non-nil resp and non-nil err, so if user opt to use it they need to handle the draining of the response body even if err is non-nil.

Reply:
I added a comment. The behavior around drain and close could be good reasons to a bigger interface that actually wraps the HTTP requests instead of being applied on the transport layer which would allow us to take care of this for the user. At the same time, even with just the standard library the user needs to make sure to adhere to practices like closing the response body when using the standard `http` package.

Reply:
having a middleware controlling the whole lifecycle of the response will likely be too much and anti-pattern. I don't think that will work :(

one alternative is that we just do `DrainAndClose(resp.Body)` if `ClientErrorFromResponse(resp)` returns non-nil, but that also means the caller would only know the status code, without knowing any body and other header in the response.

oh and also, this line should be:

```suggestion
			if err == nil {
				err = ClientErrorFromResponse(resp)
			}
```